### PR TITLE
[PY] feat: SSO-token-exchange

### PR DESCRIPTION
**Pull request by BMS-geodev**
## Linked issues

closes: #1870 

## Details

Focuses on token concepts as relates to SSO token exchange, as started in the py-sso branch started by @lilyydu 
https://github.com/microsoft/teams-ai/tree/lilyydu/py-sso. Adds basic exchange concepts to bring the py-sso branch somewhat in line with other teams-ai implementations on this topic. Also starts a few tests on the topic for later enhancement.

#### Change details

**code snippets**:

**screenshots**:

## Attestation Checklist

- [x] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (updating the doc strings in the code is sufficient)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### Additional information

> Feel free to add other relevant information below

### CODE REVIEW

File: python/packages/ai/tests/auth/test_sso_auth.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+from botbuilder.core import TurnContext
+from msal import ConfidentialClientApplication
+from botbuilder.schema import Activity, ActivityTypes, ConversationAccount, ChannelAccount
+from botbuilder.schema.teams import TeamsChannelAccount
+
+from teams.state import TurnState
+from teams.auth import SsoOptions, ConfidentialClientApplicationOptions
+from teams.auth import SsoAuth
+
+class TestSsoAuthTokenExchange(unittest.IsolatedAsyncioTestCase):
```
Comment (lilyydu) [inline python/packages/ai/tests/auth/test_sso_auth.py:13]: title of test usually matches name of file (e.g., `TestSsoAuth`) to allow for easier tracing

**Review by lilyydu — COMMENTED** [review]
just small nitpick with title, otherwise looks good

**Comment by BMS-geodev** [comment]
OK so, much more familiar with SSO and tokens now. 

I see that the primary "exchange" behaviors occur in sso_prompt.py. 

Such as acquire_token_from_cache(). 

Now, this goes through sso_dialog.py as the route to sso.py. (as an init'ed attribute in the SsoAuth class.)

sso_prompt.py has several check style methods such as _recognize_token() , and the concepts around caching. 

My question @lilyydu is: Do we need/want to bring any other "checking / verifying" steps to sso.py itself to guard the token acquisition or are needs met?

**Review by lilyydu — APPROVED** [review]

**Comment by BMS-geodev** [comment]
I believe this is ready for review and closeout. Token exchange may need to be revisited when we consider bots, but there's at least some good coverage now.

**Review by lilyydu — APPROVED** [review]

**Comment by BMS-geodev** [comment]
interesting, the lint error here is not related to the changed files:

```

==================================== ERRORS ====================================
________ ERROR collecting tests/ai/planners/test_assistants_planner.py _________
tests/ai/planners/test_assistants_planner.py:535: in <module>
    class TestAssistantsPlanner(IsolatedAsyncioTestCase):
tests/ai/planners/test_assistants_planner.py:650: in TestAssistantsPlanner
    Run(
E   pydantic_core._pydantic_core.ValidationError: 1 validation error for Run
E   parallel_tool_calls
E     Field required [type=missing, input_value={'id': '234', 'thread_id'...hread.run', 'tools': []}, input_type=dict]
E       For further information visit https://errors.pydantic.dev/2.8/v/missing
=============================== warnings summary ===============================
.venv/lib/python3.11/site-packages/recognizers_number/number/extractors.py:37
  /home/runner/work/teams-ai/teams-ai/python/packages/ai/.venv/lib/python3.11/site-packages/recognizers_number/number/extractors.py:37: SyntaxWarning: "is" with a literal. Did you mean "=="?
    if source is None or len(source.strip()) is 0:

.venv/lib/python3.11/site-packages/recognizers_date_time/date_time/utilities.py:552
  /home/runner/work/teams-ai/teams-ai/python/packages/ai/.venv/lib/python3.11/site-packages/recognizers_date_time/date_time/utilities.py:552: SyntaxWarning: 'int' object is not subscriptable; perhaps you missed a comma?
    bias = 0[0]

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.11.9-final-0 -----------
Coverage HTML written to dir coverage/html
Coverage LCOV written to file coverage/lcov.info

!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
========================= 2 warnings, 1 error in 7.13s =========================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/runner/work/teams-ai/teams-ai/python/packages/ai/scripts/test.py", line 10, in test
    subprocess.run(["poetry", "run", "pytest"], check=True)
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['poetry', 'run', 'pytest']' returned non-zero exit status 2.
Traceback (most recent call last):
------ Package[ai] ------
  File "/home/runner/work/teams-ai/teams-ai/python/scripts/test.py", line 13, in <module>
    subprocess.run(["poetry", "run", "test"], cwd=e.absolute(), check=True)
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['poetry', 'run', 'test']' returned non-zero exit status 1.
```

**Comment by BMS-geodev** [comment]
I'm aware of the TestAssistantsPlanner.py fix, its unrelated to this work, but reflects that py-sso may be out of date with main. 

Shall I ignore the lint flow error, or add the TestAssistantsPlanner.py fix `parallel_tool_calls=True` to this PR?

**Comment by BMS-geodev** [comment]
testing updated branch now.

**Review by singhk97 — APPROVED** [review]

File: python/packages/ai/teams/auth/sso/sso_dialog.py [hunk 2]
```diff
@@ -42,16 +44,15 @@ def __init__(self, name: str, options: SsoOptions, msal: ConfidentialClientAppli
         self._options = options
         self.initial_dialog_id = SSO_DIALOG_ID
 
-        # TODO - duplicate token exchange afterturn state 
-
+        # TODO - duplicate token exchange afterturn state
```
Comment (lilyydu) [inline python/packages/ai/teams/auth/sso/sso_dialog.py:47]: do we need to remove this TODO?

Comment (BMS-geodev) [reply python/packages/ai/teams/auth/sso/sso_dialog.py:47]: duplicate token exchange is handled on another PR, this TODO predated me. I think we are good to delete it as I can confirm I am actively working on that very file on this PR https://github.com/microsoft/teams-ai/pull/1890

will remove now.

**Review by lilyydu — APPROVED** [review]

**Pull request merged**